File: android/core/src/main/kotlin/in/nann/smashhockey/core/season/CareerRules.kt
package `in`.nann.smashhockey.core.season

import `in`.nann.smashhockey.core.generated.BoardRecord
import `in`.nann.smashhockey.core.generated.Career
import `in`.nann.smashhockey.core.generated.CareerRecord
import `in`.nann.smashhockey.core.generated.Club
import `in`.nann.smashhockey.core.generated.Formation
import `in`.nann.smashhockey.core.generated.Tactics
import `in`.nann.smashhockey.core.generated.TeamKey
import `in`.nann.smashhockey.core.generated.Tuning
import `in`.nann.smashhockey.core.generated.World
import java.text.Normalizer

/** A team the player is creating (spec §2.2): what the create screen edits before confirming. */
data class TeamDraft(val name: String, val short: String, val primary: Int, val secondary: Int, val world: World)

/** Why a draft cannot be confirmed (spec §2.2). Listed in this order by [CreatedTeamRules.issues]; keys match iOS. */
enum class TeamIssue(val key: String) {
    NAME_TOO_SHORT("nameTooShort"),
    NAME_TOO_LONG("nameTooLong"),
    SHORT_CODE_NOT_THREE_LETTERS("shortCodeNotThreeLetters"),
    SHORT_CODE_IS_A_CLUBS("shortCodeIsAClubs"),
    PRIMARY_NOT_IN_PALETTE("primaryNotInPalette"),
    SECONDARY_NOT_IN_PALETTE("secondaryNotInPalette"),
}

/** What the game refuses to do (spec §2.2, §11, §8.7). Every refusal is typed; none is a string. */
sealed interface GameError {
    /** A career exists: there is no switching (§2.2) — only starting over. */
    data object CareerExists : GameError
    data object NoCareer : GameError
    data class InvalidTeam(val issues: List<TeamIssue>) : GameError
    /** A new season starts only when there is none or the last one is over (§11.4). */
    data object SeasonInProgress : GameError
    data object NoSeason : GameError
    data object SeasonFinished : GameError
    /** The current matchday has no fixture for the player (they are never left on one: §11.2). */
    data object NoPlayerFixture : GameError
    data object NegativeGoals : GameError
    /** A cup match always has a winner (§8.4). */
    data object CupScoreLevel : GameError
    data object OvertimeOutsideCup : GameError
    /** Sudden death ends on the next goal (§8.4). */
    data object OvertimeNotByOneGoal : GameError
    /** A coach's board value outside §12's ranges and steps. */
    data object NotABoardValue : GameError
}

class GameException(val error: GameError) : Exception(error.toString())

internal fun refuse(error: GameError): Nothing = throw GameException(error)

/** The created team's rules (spec §2.2). */
object CreatedTeamRules {
    /** The name as it is kept: leading and trailing spaces (U+0020) dropped. */
    fun trimmedName(name: String): String = name.trim(' ')

    /**
     * The short code derived from a name: its letters A–Z (accents dropped, uppercased) — the
     * first, the second and the first later one that makes a code no club has; failing that, the
     * first two letters (as many as there are) padded with [Career.shortCodePad].
     */
    fun suggestedShortCode(name: String): String {
        val letters = Normalizer.normalize(name, Normalizer.Form.NFD).mapNotNull {
            when (it) {
                in 'A'..'Z' -> it
                in 'a'..'z' -> it.uppercaseChar()
                else -> null
            }
        }
        val clubCodes = Club.entries.map { it.short }.toSet()
        if (letters.size >= Career.shortCodeLength) {
            for (k in 2 until letters.size) {
                val code = "${letters[0]}${letters[1]}${letters[k]}"
                if (code !in clubCodes) return code
            }
        }
        return letters.take(2).joinToString("").padEnd(Career.shortCodeLength, Career.shortCodePad[0])
    }

    /** Every rule the draft breaks, in [TeamIssue] order; empty when it can be confirmed. */
    fun issues(draft: TeamDraft): List<TeamIssue> {
        val out = ArrayList<TeamIssue>()
        val name = trimmedName(draft.name)
        val length = name.codePointCount(0, name.length)
        if (length < Career.nameMinLength) out.add(TeamIssue.NAME_TOO_SHORT)
        if (length > Career.nameMaxLength) out.add(TeamIssue.NAME_TOO_LONG)
        if (draft.short.length != Career.shortCodeLength || !draft.short.all { it in 'A'..'Z' }) {
            out.add(TeamIssue.SHORT_CODE_NOT_THREE_LETTERS)
        } else if (Club.entries.any { it.short == draft.short }) {
            out.add(TeamIssue.SHORT_CODE_IS_A_CLUBS)
        }
        if (Career.kitPalette.none { it.primary == draft.primary }) out.add(TeamIssue.PRIMARY_NOT_IN_PALETTE)
        if (Career.kitPalette.none { it.secondary == draft.secondary }) out.add(TeamIssue.SECONDARY_NOT_IN_PALETTE)
        return out
    }
}

/** The player's team: always the one they created (§2.2) — there is no picking a club. */
val CareerRecord.team: TeamKey get() = TeamKey.CREATED

/**
 * The league's eight teams in their canonical order (§11.1): the clubs as declared, the player's
 * team in the place of the club it replaces.
 */
val CareerRecord.league: List<TeamKey>
    get() = Club.entries.map { club -> if (club == Career.createdReplaces) TeamKey.CREATED else TeamKey.of(club) }

/** The created team as a draft, for the screen that edits it (§16.1). */
val CareerRecord.draft: TeamDraft
    get() = TeamDraft(created.name, created.short, created.primary, created.secondary, created.world)

/** A team's short code — the player's team's own, or a club's. */
fun CareerRecord.short(of: TeamKey): String = of.club?.short ?: created.short

/** A team's rating (§2.1); the player's team's is fixed (§2.2). */
fun CareerRecord.rating(of: TeamKey): Int = of.club?.rating ?: Career.createdRating

/** The coach's board (§12): its defaults and the tactics a created team starts it from. */
object Board {
    val defaults: BoardRecord = BoardRecord(
        pressing = Tactics.defaults.pressing, covering = Tactics.defaults.covering, pushUp = Tactics.defaults.pushUp,
        discipline = Tactics.defaults.discipline, formation = Formation.entries[0],
        periodSeconds = Tuning.Board.periodSecondsDefault, ballSpinSeconds = Tuning.Board.ballSpinSecondsDefault,
    )

}

/**
 * The board with [t]'s tactics — creating a team starts the board from its tactics (§2.2) and
 * keeps the formation, period length and ball spin the player set.
 */
fun BoardRecord.withTactics(t: Tactics): BoardRecord =
    copy(pressing = t.pressing, covering = t.covering, pushUp = t.pushUp, discipline = t.discipline)

File: ios/SmashCore/Sources/SmashCore/Season/CareerRules.swift
import Foundation

/// A team the player is creating (spec §2.2): what the create screen edits before confirming.
public struct TeamDraft: Sendable, Hashable {
    public var name: String
    public var short: String
    public var primary: UInt32
    public var secondary: UInt32
    public var world: World

    public init(name: String, short: String, primary: UInt32, secondary: UInt32, world: World) {
        self.name = name
        self.short = short
        self.primary = primary
        self.secondary = secondary
        self.world = world
    }
}

/// Why a draft cannot be confirmed (spec §2.2). Listed in this order by `CreatedTeamRules.issues`.
public enum TeamIssue: String, Sendable, CaseIterable, Hashable {
    case nameTooShort
    case nameTooLong
    case shortCodeNotThreeLetters
    case shortCodeIsAClubs
    case primaryNotInPalette
    case secondaryNotInPalette
}

/// What the game refuses to do (spec §2.2, §11, §8.7). Every refusal is typed; none is a string.
public enum GameError: Error, Sendable, Equatable {
    /// A career exists: there is no switching (§2.2) — only starting over.
    case careerExists
    case noCareer
    case invalidTeam([TeamIssue])
    /// A new season starts only when there is none or the last one is over (§11.4).
    case seasonInProgress
    case noSeason
    case seasonFinished
    /// The current matchday has no fixture for the player (they are never left on one: §11.2).
    case noPlayerFixture
    case negativeGoals
    /// A cup match always has a winner (§8.4).
    case cupScoreLevel
    case overtimeOutsideCup
    /// Sudden death ends on the next goal (§8.4).
    case overtimeNotByOneGoal
    /// A coach's board value outside §12's ranges and steps.
    case notABoardValue
}

/// The created team's rules (spec §2.2).
public enum CreatedTeamRules {
    /// The name as it is kept: leading and trailing spaces (U+0020) dropped.
    public static func trimmedName(_ name: String) -> String {
        let s = Array(name.unicodeScalars)
        var lo = 0, hi = s.count
        while lo < hi, s[lo] == " " { lo += 1 }
        while hi > lo, s[hi - 1] == " " { hi -= 1 }
        var view = String.UnicodeScalarView()
        view.append(contentsOf: s[lo..<hi])
        return String(view)
    }

    /// The short code derived from a name: its letters A–Z (accents dropped, uppercased) — the
    /// first, the second and the first later one that makes a code no club has; failing that, the
    /// first two letters (as many as there are) padded with `Career.shortCodePad`.
    public static func suggestedShortCode(for name: String) -> String {
        let letters = name.decomposedStringWithCanonicalMapping.unicodeScalars.compactMap { s -> Character? in
            switch s.value {
            case 0x41...0x5A: Character(s)
            case 0x61...0x7A: Character(Unicode.Scalar(s.value - 0x20)!)
            default: nil
            }
        }
        let clubCodes = Set(Club.allCases.map(\.short))
        if letters.count >= Career.shortCodeLength {
            for k in 2..<letters.count {
                let code = String([letters[0], letters[1], letters[k]])
                if !clubCodes.contains(code) { return code }
            }
        }
        var code = String(letters.prefix(2))
        while code.count < Career.shortCodeLength { code += Career.shortCodePad }
        return code
    }

    /// Every rule the draft breaks, in `TeamIssue` order; empty when it can be confirmed.
    public static func issues(_ draft: TeamDraft) -> [TeamIssue] {
        var out: [TeamIssue] = []
        let length = trimmedName(draft.name).unicodeScalars.count
        if length < Career.nameMinLength { out.append(.nameTooShort) }
        if length > Career.nameMaxLength { out.append(.nameTooLong) }
        let short = draft.short.unicodeScalars
        if short.count != Career.shortCodeLength || !short.allSatisfy({ (0x41...0x5A).contains($0.value) }) {
            out.append(.shortCodeNotThreeLetters)
        } else if Club.allCases.contains(where: { $0.short == draft.short }) {
            out.append(.shortCodeIsAClubs)
        }
        if !Career.kitPalette.contains(where: { $0.primary == draft.primary }) { out.append(.primaryNotInPalette) }
        if !Career.kitPalette.contains(where: { $0.secondary == draft.secondary }) { out.append(.secondaryNotInPalette) }
        return out
    }
}

extension CareerRecord {
    /// The player's team: always the one they created (§2.2) — there is no picking a club.
    public var team: TeamKey { .created }

    /// The league's eight teams in their canonical order (§11.1): the clubs as declared, the
    /// player's team in the place of the club it replaces.
    public var league: [TeamKey] {
        Club.allCases.map { club in club == Career.createdReplaces ? .created : TeamKey(club) }
    }

    /// A team's short code — the player's team's own, or a club's.
    public func short(of key: TeamKey) -> String {
        key.club?.short ?? created.short
    }

    /// A team's rating (§2.1); the player's team's is fixed (§2.2).
    public func rating(of key: TeamKey) -> Int {
        key.club?.rating ?? Career.createdRating
    }
}

extension BoardRecord {
    /// The board's defaults (§12).
    public static let defaults = BoardRecord(
        pressing: Tactics.defaults.pressing, covering: Tactics.defaults.covering, pushUp: Tactics.defaults.pushUp,
        discipline: Tactics.defaults.discipline, formation: Formation.allCases[0],
        periodSeconds: Tuning.Board.periodSecondsDefault, ballSpinSeconds: Tuning.Board.ballSpinSecondsDefault)

    /// Sets the board's tactics — creating a team starts the board from its tactics (§2.2)
    /// and keeps the formation, period length and ball spin the player set.
    public mutating func adoptTactics(_ t: Tactics) {
        pressing = t.pressing
        covering = t.covering
        pushUp = t.pushUp
        discipline = t.discipline
    }
}
